sharding/shard.go

```diff
@@ -0,0 +1,219 @@
+	availabilityVal, err := s.shardDB.Get(key)
+	if err != nil {
+		return false, fmt.Errorf("key not found: %v", key)
+	}
+	var availability int
+	if err := rlp.DecodeBytes(availabilityVal, &availability); err != nil {
+		return false, fmt.Errorf("cannot RLP decode availability: %v", err)
+	}
+	if availability != 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// SetAvailability saves the availability of the chunk root in the shardDB.
+func (s *Shard) SetAvailability(chunkRoot *common.Hash, availability bool) error {
+	key := dataAvailabilityLookupKey(chunkRoot)
+	if availability {
+		enc, err := rlp.EncodeToBytes(true)
+		if err != nil {
+			return fmt.Errorf("cannot RLP encode availability: %v", err)
+		}
+		s.shardDB.Put(key, enc)
+	} else {
+		enc, err := rlp.EncodeToBytes(false)
+		if err != nil {
+			return fmt.Errorf("cannot RLP encode availability: %v", err)
+		}
+		s.shardDB.Put(key, enc)
+	}
+	return nil
+}
+
+// SaveHeader adds the collation header to shardDB.
+func (s *Shard) SaveHeader(header *CollationHeader) error {
+	encoded, err := header.EncodeRLP()
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %v", err)
+	}
+
+	// Uses the hash of the header as the key.
+	hash := header.Hash()
+	s.shardDB.Put(hash, encoded)
+	return nil
+}
+
+// SaveBody adds the collation body to the shardDB and sets availability.
+func (s *Shard) SaveBody(body []byte) error {
+	// TODO: check if body is empty and throw error.
+	// TODO: dependent on blob serialization.
+	// chunkRoot := getChunkRoot(body) using the blob algorithm utils.
+	// right now we will just take the raw keccak256 of the body until #92 is merged.
+	chunkRoot := common.BytesToHash(body)
+	s.shardDB.Put(chunkRoot, body)
+	s.SetAvailability(&chunkRoot, true)
+	return nil
+}
+
+// SaveCollation adds the collation's header and body to shardDB.
+func (s *Shard) SaveCollation(collation *Collation) error {
+	if err := s.ValidateShardID(collation.Header()); err != nil {
+		return err
+	}
+	s.SaveHeader(collation.Header())
+	s.SaveBody(collation.Body())
+	return nil
+}
+
+// SetCanonical sets the collation as canonical in the shardDB. This is called
+// after the period is over and over 2/3 notaries voted on the header.
+func (s *Shard) SetCanonical(header *CollationHeader) error {
+	if err := s.ValidateShardID(header); err != nil {
+		return err
+	}
+	// the header needs to have been stored in the DB previously, so we
+	// fetch it from the shardDB.
+	hash := header.Hash()
+	dbHeader, err := s.HeaderByHash(&hash)
+	if err != nil {
+		return err
+	}
+	key := canonicalCollationLookupKey(dbHeader.ShardID(), dbHeader.Period())
+	encoded, err := dbHeader.EncodeRLP()
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %v", err)
+	}
+	s.shardDB.Put(key, encoded)
+	return nil
+}
+
+// dataAvailabilityLookupKey formats a string that will become a lookup
+// key in the shardDB.
+func dataAvailabilityLookupKey(chunkRoot *common.Hash) common.Hash {
```

Comment:
Why do you need to add a prefix to the hash then hash it again? Can't the lookup key for a chunkRoot be the hash itself?

Reply:
Nevermind.